Lazy Loading of Subgraphs for Workflow Instance Viewer

sfmskywalker opened the pull request:
This PR implements lazy loading of workflow activities. This allows the backend to provide a layered subgraph model instead of a fully resolved graph which would require serializing all workflow definitions referenced via workflow activities.

Closes https://github.com/elsa-workflows/elsa-core/issues/5350

sfmskywalker committed:
Remove 'includeCompositeRoot' parameter in workflow services

This commit removes the 'includeCompositeRoot' parameter from calls to FindByDefinitionIdAsync, FindByIdAsync, and FindManyByIdAsync methods in RemoteWorkflowDefinitionService and IWorkflowDefinitionService. It also changes the Elsa.Api.Client reference from package to project in Elsa.Studio.Core project file.

sfmskywalker committed:
Update tool and package versions to 3.3.0

The tool version has been updated in ToolVersion.cs. This change is reflected in the base_version environment variable in the GitHub actions configuration file. Additionally, the version in the GitHub actions file is no longer hardcoded but now references this base_version variable, ensuring consistency and easier future updates.

sfmskywalker committed:
Merge branch 'patch/3.2.x'

sfmskywalker committed:
Merge branch 'main' into enhancement/lazy-loading-sub-workflows

sfmskywalker committed:
Remove flag in workflow definition lookup

sfmskywalker committed:
Merge branch 'patch/3.2.x'

sfmskywalker committed:
Merge branch 'main' into perf/5350

sfmskywalker committed:
Update WorkflowInstanceViewer and related components for subgraph support

Added support for subgraph in WorkflowInstanceViewer and related components. Made necessary changes in various classes including FallbackDesignerProvider, WorkflowInstanceDesigner, and others to incorporate the support for subgraph. The modifications were also reflected in the IDiagramDesignerProvider interface and IDiagramDesignerService to accept a WorkflowSubgraph parameter. Adjusted PathChanged event handling accordingly.

sfmskywalker committed:
Add IgnoreChildActivitiesConverter and improve activity JSON conversion

A new JSON converter, IgnoreChildActivitiesConverter, has been introduced to handle specific serialization scenarios. On top of this, improvements have been made to the existing ActivityJsonConverter. This includes more robust handling of synthetic properties and the ability to ignore child activities during serialization. In DiagramDesignerWrapper, an addition has been made to facilitate the handling of embedded activities.

sfmskywalker committed:
Remove WorkflowSubgraph from various components and services

This commit removes the WorkflowSubgraph from multiple components and services across the project. It includes adjustments made to various files to replace the usage of WorkflowSubgraph. This has resulted in fewer parameters in the activity functions and cleaner code. It's a key revision to streamline the codebase and simplify the future development process.

sfmskywalker committed:
Refactor and enhance AddNewActivityAsync method

Refactored the AddNewActivityAsync method in FlowchartDesignerWrapper.razor.cs to improve readability and also enhanced the method by setting a new 'nodeId' for the added activity, making the activity aware of its parent and its own identity. This addition helps in better handling and identification of activity in the flowchart.

sfmskywalker committed:
Update path to Elsa.Api.Client project

The path to 'Elsa.Api.Client' in the solution file has been updated. It now points to the 'perf-5350' folder instead of 'main', to reflect its new location in the codebase.

sfmskywalker committed:
Update project references and port configurations

The commit includes changes in various files that primarily update the project reference of Elsa.Api.Client.csproj to a different branch and modify ASPNETCORE_URLS and Backend server URLs. Additionally, better variable naming for enhanced code readability and logic adjustments are implemented in the DiagramDesignerWrapper.razor.cs file.

sfmskywalker committed:
Update Backend URL in appsettings.json

In the Elsa Studio Host, the Backend URL in the appsettings.json file has been changed. The port number was updated from 5001 to 8001 in the URL, adjusting the endpoint that the application communicates with.

sfmskywalker committed:
Update Elsa.Api.Client package to version 3.2.0-rc2.1823

The Elsa.Api.Client package has been updated to version 3.2.0-rc2.1823, which is referenced directly instead of the earlier project reference. This change applies to the Elsa.Studio.Core project and the package version property.

sfmskywalker committed:
Update server ports, tool version and clean up code

The commit includes changes to the server ports specified in different configuration files. The version of the tool has been downgraded from 3.3.0 to 3.2.0 across multiple files. Unnecessary code, mainly concerning an unused tab index and panel change listener, has been removed from WorkflowInstanceWorkspace.

Unnecessary mentions of the `Elsa.Api.Client.Resources.WorkflowDefinitions.Models` namespace are also removed from various files. Other unnecessary redundancies, like the OnActivePanelIndexChanged function, are also omitted.

sfmskywalker committed:
Refactor StitchNodesRecursive method in DiagramDesignerWrapper

The method StitchNodesRecursive in DiagramDesignerWrapper has been refactored for better readability and clarity. A new variable `isCollection` was introduced to simplify the conditional assignment. Also, unnecessary curly brackets surrounding the loop body were removed to streamline the code.

raymonddenhaan commented on Elsa.Studio-Core.sln:
```diff
@@ -83,7 +83,7 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "BlazorApp1", "samples\Blazo
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "client", "client", "{F4B408E1-5169-4FA9-B2D7-2377634C4DEE}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Elsa.Api.Client", "..\..\elsa-core\main\src\clients\Elsa.Api.Client\Elsa.Api.Client.csproj", "{26ECE061-0F39-4945-8356-EF14E67A8F39}"
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Elsa.Api.Client", "..\..\elsa-core\perf-5350\src\clients\Elsa.Api.Client\Elsa.Api.Client.csproj", "{26ECE061-0F39-4945-8356-EF14E67A8F39}"
```
Reminder to revert this before merging

raymonddenhaan commented on src/framework/Elsa.Studio.Core/Elsa.Studio.Core.csproj:
```diff
@@ -33,9 +33,9 @@
-<!--        <ProjectReference Include="..\..\..\..\..\elsa-core\main\src\clients\Elsa.Api.Client\Elsa.Api.Client.csproj"/>-->
-<!--    </ItemGroup>-->
+    <!--    <ItemGroup Label="Elsa">-->
+    <!--        <ProjectReference Include="..\..\..\..\..\elsa-core\perf-5350\src\clients\Elsa.Api.Client\Elsa.Api.Client.csproj"/>-->
```
Reminder to revert this before merging

raymonddenhaan commented on src/modules/Elsa.Studio.Workflows/DiagramDesigners/Flowcharts/FlowchartDesignerWrapper.razor.cs:
```diff
@@ -23,38 +23,45 @@ public partial class FlowchartDesignerWrapper
     /// </summary>
-    [Parameter] public JsonObject Flowchart { get; set; } = default!;
-    
+    [Parameter]
```
In the `WorkflowInstanceViewer.razor.cs` the attributes are on the same line. I think it is good to use 1 style for this.

raymonddenhaan commented on src/modules/Elsa.Studio.Workflows/Shared/Components/DiagramDesignerWrapper.razor.cs:
```diff
@@ -126,7 +129,7 @@ public async Task SelectActivityAsync(ActivityNode activityNode)
                 break;
 
-            var propName = flowchart!.PropertyName!.Pascalize();
+            var propName = flowchart!.Port!.Pascalize();
```
Let's rename the variable as well

raymonddenhaan commented on src/modules/Elsa.Studio.Workflows/Shared/Components/DiagramDesignerWrapper.razor.cs:
```diff
@@ -347,6 +368,33 @@ private async Task OnActivityEmbeddedPortSelected(ActivityEmbeddedPortSelectedAr
+                    var parentNodeId = activity.GetNodeId();
+                    var subGraph = await WorkflowDefinitionService.FindSubgraphAsync(WorkflowDefinitionVersionId, parentNodeId);
+
+                    if (subGraph != null)
```
Should we not add some kind of feedback to the user that we couldn't find (load) the workflow if this is null?

sfmskywalker replied:
Yes, in fact we should throw an exception because this would only happen if there's a serialization bug.

sfmskywalker committed:
Add explanation for StitchNodesRecursive function

The new code provides an illustrative example of how the StitchNodesRecursive function operates in the file DiagramDesignerWrapper.razor.cs. It makes clear how the process of converting a graph model of nodes to a structured model where child activities are properties of parent properties, works.

sfmskywalker committed:
Refine comment in DiagramDesignerWrapper

This commit makes minor modification to a comment in the DiagramDesignerWrapper file in Elsa Studio Workflows. The update clarifies that child activities are properties of their specific parent activity, rather than generic parent properties.

sfmskywalker committed:
Refactor code formatting and improve error handling

Adjusted code clustering and exception handling in workflow modules. Main changes include adjusting code layout for better readability and enhancing error management in activity subgraph selection to prevent silent failures. This will improve code maintainability and robustness.

sfmskywalker committed:
Refactor StitchNodesRecursive method call

The StitchNodesRecursive method call and embeddedActivity assignment in the DiagramDesignerWrapper class are updated. They now work with selectedPortGraph instead of selectedActivityGraph. This change ensures the accurate assignment of the port to the specific activity graph.

sfmskywalker committed:
Refactor exception handling in DiagramDesignerWrapper

Exceptions for missing activity and port graphs in the method DiagramDesignerWrapper are refactored to clean up the code. Now, these exceptions are thrown directly within the assignment lines for both selectedActivityGraph and selectedPortGraph, reducing the number of lines and improving readability.

sfmskywalker committed:
Update Elsa.Studio-Core settings

The .DotSettings file of Elsa.Studio-Core solution has been updated. It now includes additional ReSharper settings related to code style and formatting. This should help maintain consistent coding styles and improve the readability of the code.

sfmskywalker committed:
Remove StitchNodesRecursive method

The StitchNodesRecursive method that was used to convert a graph model of nodes to a structured model has been removed. As part of the code restructuring, it was no longer needed. We now directly access the required items from the 'selectedActivityGraph' object. This simplifies the codebase and reduces complexity.

sfmskywalker committed:
Update Elsa.Api.Client package version

The version of the Elsa.Api.Client package has been updated from 3.2.0-rc2.1823 to 3.2.0-rc2.1834 to include the most recent changes and improvements. This update ensures that the application is making use of the latest version of the package.

sfmskywalker committed:
Update project references in solution files

This commits updates the project references across various solution files. It specifically replaces references from the "perf-5350" branch to the "main" branch in the Elsa.Studio.Core project.

sfmskywalker committed:
Merge remote-tracking branch 'origin/patch/3.2.x' into perf/5350

sfmskywalker committed:
Remove duplicate IsReadOnly property definition

The IsReadOnly property was defined twice in the WorkflowDefinitionEditor component, causing redundancy. This commit removes the duplicate definition to clean up the code and avoid any potential conflicts or confusion.
